src/DiagnosticsManager.ts

```diff
@@ -0,0 +1,281 @@
+                )
+                .catch(e =>
+                    context.outputChannel.log(`${e}`, 'Failed to provide "swiftc" diagnostics')
+                );
+        });
+    }
+
+    /**
+     * Provide a new list of diagnostics for a given file
+     *
+     * @param uri {@link vscode.Uri Uri} of the file these diagonstics apply to
+     * @param sources The source of the diagnostics which will apply for cleaning
+     * up diagnostics that have been removed. See {@link swiftc} and {@link sourcekit}
+     * @param diagnostics Array of {@link vscode.Diagnostic}. This can be empty to remove old diagnostics for the specified `sources`.
+     */
+    handleDiagnostics(uri: vscode.Uri, sources: string[], diagnostics: vscode.Diagnostic[]): void {
+        // Remove the old set of benefits from this source
+        const currentDiagnostics = this.removeDiagnostics(
+            this.diagnosticCollection.get(uri)?.slice() || [],
+            sources
+        );
+        let newDiagnostics: vscode.Diagnostic[];
+        switch (configuration.diagnosticsCollection) {
+            case "keepSourceKit":
+                newDiagnostics = this.mergeDiagnostics(
+                    currentDiagnostics,
+                    diagnostics,
+                    DiagnosticsManager.sourcekit
+                );
+                break;
+            case "keepSwiftc":
+                newDiagnostics = this.mergeDiagnostics(
+                    currentDiagnostics,
+                    diagnostics,
+                    DiagnosticsManager.swiftc
+                );
+                break;
+            case "onlySourceKit":
+                newDiagnostics = this.removeDiagnostics(
+                    currentDiagnostics,
+                    DiagnosticsManager.swiftc
+                ); // Just in case
+                if (DiagnosticsManager.swiftc.find(s => sources.includes(s))) {
+                    break;
+                }
+                newDiagnostics.push(...diagnostics);
+                break;
+            case "onlySwiftc":
+                newDiagnostics = this.removeDiagnostics(
+                    currentDiagnostics,
+                    DiagnosticsManager.sourcekit
+                ); // Just in case
+                if (DiagnosticsManager.sourcekit.find(s => sources.includes(s))) {
+                    break;
+                }
+                newDiagnostics.push(...diagnostics);
+                break;
+            case "keepAll":
+                newDiagnostics = currentDiagnostics.concat(diagnostics);
+                break;
+        }
+        this.diagnosticCollection.set(uri, newDiagnostics);
+    }
+
+    private mergeDiagnostics(
+        currentDiagnostics: vscode.Diagnostic[],
+        newDiagnostics: vscode.Diagnostic[],
+        precedence: string[]
+    ): vscode.Diagnostic[] {
+        const merged: vscode.Diagnostic[] = [];
+        const messages: string[] = [];
+        for (const diagnostic of newDiagnostics) {
```

Comment:
Is it not possible to do this just by merging straight  into `currentDiagnostics` instead 
- Creating a new list of unique build diagnostics
- Then creating a list from that of the ones that need to be merged into `currentDiagnostics`
- Then returning the union of that list with `currentDiagnostics`

Reply:
Done